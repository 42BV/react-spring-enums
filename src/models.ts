export type ComplexEnumType = {
  code: string;
  displayName: string;
};

export type EnumValues = string[] | ComplexEnumType[];
export type EnumValue = string | ComplexEnumType;

/**
 * The enums should have the following signature for simple enums:
 *
 * {
 *   "UserRole": [
 *     "ADMIN",
 *     "USER"
 *  ],
 *  "BillingInterval": [
 *    "MONTHLY", "WEEKLY", "DAILY"
 *  ]
 * }
 *
 * or for complex enums:
 *
 * {
 *   "UserRole": [
 *     {
 *       "code": "ADMIN",
 *       "displayName": "Administrator"
 *     },
 *     {
 *       "code": "USER",
 *       "displayName": "User"
 *     }
 *   ],
 *   "BillingInterval": [
 *     {
 *       "code": "MONTHLY",
 *       "displayName": "Monthly"
 *     },
 *     {
 *       "code": "WEEKLY",
 *       "displayName": "Weekly",
 *       "onlyFor": "SPECIAL_CUSTOMER"
 *     }
 *   ]
 * }
 *
 * The keys represent the name of the enum, in the above case 'UserRole',
 * and 'BillingInterval'. The array points to the possible values
 * of the enum. In case of the complex enums, the code represents the value
 * and the displayName can be used to display i.e. a translation without
 * having to integrate a translation library or fetch a translation.
 */
export type Enums = {
  [enumName: string]: EnumValues;
};
