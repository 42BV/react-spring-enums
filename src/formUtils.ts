import { Page, pageOf } from '@42.nl/spring-connect';
import { EnumValue } from './models';

/**
 * Returns the enumValues matching the provided query.
 *
 * @export
 * @param {EnumValue[]} enumValues
 * @param {string} query
 * @param {Function} getDisplayValue Function that accepts an enum value and returns a string to match with the query
 * @returns {string[]}
 * @throws Throws an error if enumValues is not an array of type
 * EnumValue (string or object with displayName property), and
 * getDisplayValue is not provided.
 */
export function filterEnumValues<T>(
  enumValues: T[],
  query?: string,
  getDisplayValue = enumValueDisplayName<T>
): T[] {
  if (query) {
    return enumValues.filter((value) =>
      getDisplayValue(value).match(new RegExp(query, 'i'))
    );
  }

  return enumValues;
}

/**
 * Takes an `EnumValue` array, page and query and turns it into
 * a `Page` from `@42.nl/spring-connect` which is based on a Spring
 * boot Page:
 *
 * https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/domain/Page.html
 *
 * The `query` will filter out any enums which do not start with that
 * query.
 *
 * The `page` is the page it is currently on, it determines the slice
 * which the user currently sees.
 *
 * The `size` defaults to 10 but can be any positive number.
 *
 * The `oneBased`, which default to `true`, determines if the page
 * should start at page 1 or page 0.
 *
 * @export
 * @param {{
 *   enumValues: EnumValue[];
 *   size?: number;
 *   page: number;
 *   query: string;
 *   oneBased?: boolean;
 * }} {
 *   enumValues,
 *   page,
 *   size = 10,
 *   query,
 *   oneBased = true
 * }
 * @returns {Promise<Page<EnumValue>>}
 * @throws Throws an error if enumValues is not an array of type
 * EnumValue (string or object with displayName property), and
 * getDisplayValue is not provided.
 */
export function getEnumsAsPage<T = EnumValue>({
  enumValues,
  page,
  size = 10,
  query,
  oneBased = true,
  getDisplayValue
}: {
  enumValues: T[];
  size?: number;
  page: number;
  query: string;
  oneBased?: boolean;
  getDisplayValue?: (enumValue: T) => string;
}): Promise<Page<T>> {
  return new Promise((resolve) =>
    resolve(
      pageOf(
        filterEnumValues<T>(enumValues, query, getDisplayValue),
        page,
        size,
        oneBased
      )
    )
  );
}

function enumValueDisplayName<T = EnumValue>(enumValue: T): string {
  if (typeof enumValue === 'string') {
    return enumValue;
  }

  if (
    enumValue &&
    typeof enumValue === 'object' &&
    enumValue.hasOwnProperty('displayName')
  ) {
    // @ts-expect-error We checked if property displayName exists but typescript somehow still doesn't understand
    return enumValue.displayName;
  }

  throw new Error(
    'You have to provide a custom getDisplayValue for getEnumsAsPage and filterEnumValues when the enum type is not a string and does not extend ComplexEnumType'
  );
}
