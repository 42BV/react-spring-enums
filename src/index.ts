export { configureEnums, Config } from './config';
export { loadEnums } from './load-enums';
export { EnumsProvider, EnumsContext } from './provider';
export { useEnums, useEnum } from './hooks';
export { Enums, EnumValues } from './models';
export { makeEnumsService, getDefaultState, EnumsState } from './service';
export { filterEnumValues, getEnumsAsPage } from './formUtils';
