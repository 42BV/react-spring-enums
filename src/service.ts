import { Enums } from './models';

export type EnumsState<T = Enums> = {
  enums: T;
};
export type Subscriber<T> = (state: EnumsState<T>) => void;

export type EnumsService<T = Enums> = {
  subscribe(subscriber: Subscriber<T>): void;
  unsubscribe(subscriber: Subscriber<T>): void;
  getState: () => EnumsState<T>;
  setEnums(enums: T): void;
};

export function getDefaultState<T>(): EnumsState<T> {
  return {
    enums: {} as T
  };
}

/**
 * Creates a EnumsService which stores Spring enums.
 *
 * Exposes a subscription to anyone who wants to listen to the
 * changes to the EnumsState.
 */
export function makeEnumsService<T = Enums>(): EnumsService<T> {
  let state = getDefaultState<T>();
  let subscribers: Subscriber<T>[] = [];

  function getState() {
    return state;
  }

  function subscribe(subscriber: Subscriber<T>): void {
    subscribers.push(subscriber);

    subscriber(state);
  }

  function informSubscribers() {
    subscribers.forEach((subscriber) => subscriber({ ...state }));
  }

  function setEnums(enums: T): void {
    state = { enums };

    informSubscribers();
  }

  function unsubscribe(subscriber: Subscriber<T>): void {
    subscribers = subscribers.filter((s) => s !== subscriber);
  }

  return {
    getState,
    subscribe,
    unsubscribe,
    setEnums
  };
}
