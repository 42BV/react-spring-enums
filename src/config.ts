import { EnumsService, makeEnumsService } from './service';
import { Enums } from './models';

export type Config = {
  // The URL which will provide the enums over a GET request.
  enumsUrl: string;
};

let config: Config | null = null;
let service: EnumsService | null = null;

/**
 * Configures the Constraint library.
 *
 * @param {Config} c The new configuration
 */
export function configureEnums(c: Config): void {
  config = c;
  service = makeEnumsService();
}

/**
 * Either returns a Config or throws an error when the
 * config is not yet initialized.
 *
 * @returns The Config
 */
export function getConfig(): Config {
  if (config === null) {
    throw new Error('The enum service is not initialized.');
  } else {
    return config;
  }
}

export function getService<T = Enums>(): EnumsService<T> {
  if (service === null) {
    throw new Error('The enum service is not initialized.');
  } else {
    return service as EnumsService<T>;
  }
}
