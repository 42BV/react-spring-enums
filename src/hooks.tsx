import { useEffect, useState } from 'react';

import { getService } from './config';
import { EnumsState } from './service';
import { Enums, EnumValue } from './models';

class MissingEnumException extends Error {}

/**
 * Retrieves the entire enum state.
 */
export function useEnums<T = Enums>(): EnumsState<T> {
  const service = getService<T>();
  const [state, setState] = useState<EnumsState<T>>(service.getState());

  useEffect(() => {
    const subscriber = (nextState: EnumsState<T>) => {
      setState(nextState);
    };

    service.subscribe(subscriber);

    return () => {
      service.unsubscribe(subscriber);
    };
  }, []);

  return state;
}

/**
 * Retrieves a single enum from the state.
 *
 * @param enumName
 */
export function useEnum<T = EnumValue>(enumName: string): T[] {
  const { enums } = useEnums<{ [key: string]: T[] }>();
  const enumValues = enums[enumName];

  if (enumValues) {
    return enumValues;
  } else {
    throw new MissingEnumException(
      `@42.nl/spring-enum: The enum named '${enumName}' could not be found, make sure the enums are loaded before the using them and that the '${enumName}' enum actually exists.`
    );
  }
}
