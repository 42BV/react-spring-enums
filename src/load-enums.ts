import { getConfig, getService } from './config';

// Throw error when not 200 otherwise parse response.
function tryParse(response: Response): Promise<any> {
  if (response.status !== 200) {
    throw response;
  } else {
    return response.json();
  }
}

/**
 * Loads the enums from the back-end.
 *
 * The URL it will send the request to is defined by the 'enumsUrl'
 * from the Config object. The HTTP method it uses is 'get'.
 *
 * It will also send the credentials if 'needsAuthentication' from
 * the Config object is set to true.
 *
 * The entire response will be written to the Redux EnumsStore's
 * enums key. Whatever the JSON response is will be the enums.
 *
 *  An example response:
 *
 * ```JSON
 * {
 *   "UserRole": [
 *     "ADMIN",
 *     "USER"
 *  ],
 *  "BillingInterval": [
 *    "MONTHLY", "WEEKLY", "DAILY"
 *  ]
 * ```
 *
 * @returns {Promise}
 */
export async function loadEnums(): Promise<void> {
  const { enumsUrl, needsAuthentication } = getConfig();
  const service = getService();

  const config: RequestInit = needsAuthentication ? { credentials: 'include' } : {};
  const response = await fetch(enumsUrl, config);
  const enums = await tryParse(response);

  service.setEnums(enums);
}
