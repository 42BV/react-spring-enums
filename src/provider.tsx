import React, { createContext } from 'react';

import { useEnums } from './hooks';
import { EnumsState } from './service';

export const EnumsContext = createContext<EnumsState>({ enums: {} });

export function EnumsProvider({ children }: { children: React.ReactNode }) {
  const enums = useEnums();
  return (
    <EnumsContext.Provider value={enums}>{children}</EnumsContext.Provider>
  );
}
