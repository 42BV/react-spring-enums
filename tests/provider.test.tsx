import React from 'react';
import { EnumsProvider, EnumsContext } from '../src/provider';
import { configureEnums, getService } from '../src/config';
import { useEnums } from '../src/hooks';
import renderer from 'react-test-renderer';

const HookTest = () => {
  const {
    enums: { CAR_TYPES },
  } = useEnums();
  return <h2>I really like {CAR_TYPES[0]} cars</h2>;
};

class ConsumerTest extends React.Component<{ favoriteCar: string }> {
  render() {
    const { favoriteCar } = this.props;
    return <h2>My favorite car brand is {favoriteCar}</h2>;
  }
}

describe('EnumProvider', () => {
  function setup({ needsAuthentication }: { needsAuthentication: boolean }): void {
    configureEnums({
      enumsUrl: '/api/enums',
      needsAuthentication,
    });

    getService().setEnums({ CAR_TYPES: ['AUDI', 'TESLA'] });
  }

  test('should provide context as hook', () => {
    setup({ needsAuthentication: false });
    const tree = renderer
      .create(
        <EnumsProvider>
          <HookTest />
        </EnumsProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should provide context as consumer', () => {
    const tree = renderer
      .create(
        <EnumsProvider>
          <EnumsContext.Consumer>
            {({ enums: { CAR_TYPES } }) => <ConsumerTest favoriteCar={CAR_TYPES[1]} />}
          </EnumsContext.Consumer>
        </EnumsProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
