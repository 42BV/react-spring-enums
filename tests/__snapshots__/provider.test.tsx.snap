// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnumProvider should provide context as consumer 1`] = `
<h2>
  My favorite car brand is
   
  TESLA
</h2>
`;

exports[`EnumProvider should provide context as hook 1`] = `
<p>
  I really like 
  AUDI
   cars
</p>
`;
